logics/impls/marketplace/marketplace_sale.rs

```diff
@@ -170,66 +170,61 @@ where
         contract_address: AccountId,
         token_id: Id,
     ) -> Result<(), MarketplaceError> {
-        if let Some(item) = self
+        let item = self
             .data::<Data>()
             .items
             .get(&(contract_address, token_id.clone()))
-        {
-            let value = Self::env().transferred_value();
-            self.check_value(value, item.price)?;
-
-            let marketplace_fee = value
-                .checked_mul(self.data::<Data>().fee as u128)
-                .unwrap_or_default()
-                .checked_div(10_000)
-                .unwrap_or_default();
-
-            let collection = self
-                .data::<Data>()
-                .registered_contracts
-                .get(&contract_address)
-                .unwrap();
-            let author_royalty = value
-                .checked_mul(collection.royalty as u128)
-                .unwrap_or_default()
-                .checked_div(10_000)
-                .unwrap_or_default();
-            let seller_fee = value
-                .checked_sub(marketplace_fee)
-                .unwrap_or_default()
-                .checked_sub(author_royalty)
-                .unwrap_or_default();
-
-            if let Some(token_owner) = PSP34Ref::owner_of(&contract_address, token_id.clone()) {
-                let caller = Self::env().caller();
-                ensure!(token_owner != caller, MarketplaceError::AlreadyOwner);
-
-                if PSP34Ref::transfer(
-                    &contract_address,
-                    caller,
-                    token_id,
-                    ink_prelude::vec::Vec::new(),
-                ) == Ok(())
-                {
-                    Self::env()
-                        .transfer(token_owner, seller_fee)
-                        .map_err(|_| MarketplaceError::TransferToOwnerFailed)?;
-                    Self::env()
-                        .transfer(self.data::<Data>().market_fee_recipient, marketplace_fee)
-                        .map_err(|_| MarketplaceError::TransferToMarketplaceFailed)?;
-                    Self::env()
-                        .transfer(collection.royalty_receiver, author_royalty)
-                        .map_err(|_| MarketplaceError::TransferToAuthorFailed)?;
-
-                    return Ok(())
-                } else {
-                    return Err(MarketplaceError::UnableToTransferToken)
-                }
-            } else {
-                return Err(MarketplaceError::NotOwner)
+            .ok_or(MarketplaceError::ItemNotListedForSale)?;
+
+        let token_owner = PSP34Ref::owner_of(&contract_address, token_id.clone())
+            .ok_or(MarketplaceError::CanNotDetermineOwner)?;
```

Comment:
I think if token has no owner that's means it doesn't exit.
[link](https://github.com/Supercolony-net/openbrush-contracts/blob/e366f6ff1e5892c6a624833dd337a6da16a06baa/contracts/src/token/psp34/psp34.rs#L183)
I will return `TokenDoesNotExist` instead.